src/scf/newton.jl

```diff
@@ -0,0 +1,410 @@
+function proj(δφ, φ; tol_test=1e-12)
+
+    Nk1 = size(δφ,1)
+    Nk2 = size(φ,1)
+    @assert Nk1 == Nk2
+    Nk = Nk1
+
+    Πδφ = similar(δφ)
+
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        Πδφk = deepcopy(δφk)
+
+        N1 = size(δφk,2)
+        N2 = size(φk,2)
+        @assert N1 == N2
+        N = N1
+
+        for i = 1:N, j = 1:N
+            Πδφk[:,i] -= (φk[:,j]'δφk[:,i]) * φk[:,j]
+        end
+        Πδφ[ik] = Πδφk
+    end
+
+    # test orthogonalisation
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        Πδφk = Πδφ[ik]
+        N = size(φk,2)
+        for i = 1:N, j = 1:N
+            @assert abs(Πδφk[:,i]'φk[:,j]) < tol_test [println(abs(Πδφk[:,i]'φk[:,j]))]
+        end
+    end
+
+    Πδφ
+end
+
+# KrylovKit custom orthogonaliser to be used in KrylovKit eigsolve, svdsolve,
+# linsolve, ...
+# This has to be passed in KrylovKit solvers when studying (Ω+K) as a super
+# operator from the tangent space to the tangent space to be sure that
+# iterations are constrained to stay on the tangent space.
+struct OrthogonalizeAndProject{F, O <: Orthogonalizer, φ} <: Orthogonalizer
+    projector::F
+    orth::O
+    φ::φ
+end
+OrthogonalizeAndProject(projector, φ) = OrthogonalizeAndProject(projector,
+                                                                KrylovDefaults.orth,
+                                                                φ)
+function KrylovKit.orthogonalize!(v::T, b::OrthonormalBasis{T}, x::AbstractVector,
+                                        alg::OrthogonalizeAndProject) where {T}
+    v, x = orthogonalize!(v, b, x, alg.orth)
+    v = reshape(v, size(alg.φ))
+    v = pack(alg.projector(v, alg.φ))::T
+    v, x
+end
+function KrylovKit.orthogonalize!(v::T, x::AbstractVector,
+                                        alg::OrthogonalizeAndProject) where {T}
+    v, x = orthogonalize!(v, x, alg.orth)
+    v = reshape(v, size(alg.φ))
+    v = pack(alg.projector(v, alg.φ))::T
+    v, x
+end
+function KrylovKit.gklrecurrence(operator, U::OrthonormalBasis, V::OrthonormalBasis, β,
+                                 alg::OrthogonalizeAndProject)
+    u = U[end]
+    v = operator(u, true)
+    v = axpy!(-β, V[end], v)
+    # for q in V # not necessary if we definitely reorthogonalize next step and previous step
+    #     v, = orthogonalize!(v, q, ModifiedGramSchmidt())
+    # end
+    α = norm(v)
+    rmul!(v, inv(α))
+
+    r = operator(v, false)
+    r = axpy!(-α, u, r)
+    for q in U
+        r, = orthogonalize!(r, q, alg)
+    end
+    β = norm(r)
+    return v, r, α, β
+end
+
+############################# OPERATORS ########################################
+
+"""
+    apply_Ω(basis::PlaneWaveBasis, δφ, φ, H, λ)
+
+Compute the application of Ω to an element in the space tangent to φ.
+It can be done analitically with the formula
+      Ωδφi = H*δφi - Σj <δφj|H|δφi> δφj - Σj λji δφj
+where λij = <φi|H|φj>;
+"""
+function apply_Ω(basis::PlaneWaveBasis, δφ, φ, H)
```

Comment:
Is this so specific to Newton that it does not make to put this in a separate file?

Reply:
It is specific to Newton in the sense that we need the Jacobian to perform the iterations. But we might need to compute Ω or K for other reasons. I wanted to put them in a separate file at the beginning but @antoine-levitt seems to prefer to have everything in one file. I don't have any opinion personnaly, both are fine to me !